runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/GreedilyFusedExecutableStage.java

```diff
@@ -0,0 +1,223 @@
+            .orElseThrow(
+                () ->
+                    new IllegalArgumentException(
+                        String.format(
+                            "%s must be populated on all %s in a %s",
+                            Environment.class.getSimpleName(),
+                            PTransformNode.class.getSimpleName(),
+                            GreedilyFusedExecutableStage.class.getSimpleName())));
+    initialNodes.forEach(
+        transformNode ->
+            checkArgument(
+                environment.equals(pipeline.getEnvironment(transformNode).orElse(null)),
+                "All %s in a %s must be the same. Got %s and %s",
+                Environment.class.getSimpleName(),
+                GreedilyFusedExecutableStage.class.getSimpleName(),
+                environment,
+                pipeline.getEnvironment(transformNode).orElse(null)));
+    this.pipeline = pipeline;
+    this.maybeInputPCollection = inputPCollection;
+    this.fusedCollections = new HashSet<>();
+    this.fusedTransforms = new LinkedHashSet<>(initialNodes);
+    this.materializedPCollections = new HashSet<>();
+
+    this.fusionCandidates = new ArrayDeque<>();
+    for (PTransformNode initialConsumer : initialNodes) {
+      fusionCandidates.addAll(pipeline.getOutputPCollections(initialConsumer));
+    }
+    while (!fusionCandidates.isEmpty()) {
+      tryFuseCandiate(fusionCandidates.poll());
+    }
+  }
+
+  private void tryFuseCandiate(PCollectionNode candidate) {
+    if (materializedPCollections.contains(candidate) || fusedCollections.contains(candidate)) {
+      // This should generally mean we get to a Flatten via multiple paths through the graph and
+      // we've already determined what to do with the output.
+      LOG.debug(
+          "Skipping fusion candidate {} because it is {} already in this {}",
+          candidate,
+          materializedPCollections.contains(candidate) ? "materialized" : "fused",
+          GreedilyFusedExecutableStage.class.getSimpleName());
+      return;
+    }
+    boolean canFuse = true;
+    for (PTransformNode node : pipeline.getPerElementConsumers(candidate)) {
+      if (!(GreedyPCollectionFusers.canFuse(node, this, pipeline))) {
+        canFuse = false;
+        break;
+      }
+    }
+    if (canFuse) {
+      // All of the consumers of the candidate PCollection can be fused into this stage, so do so.
+      fusedCollections.add(candidate);
+      fusedTransforms.addAll(pipeline.getPerElementConsumers(candidate));
+      for (PTransformNode consumer : pipeline.getPerElementConsumers(candidate)) {
+        // The outputs of every transform fused into this stage must be either materialized or
+        // themselves fused away, so add them to the set of candidates.
+        fusionCandidates.addAll(pipeline.getOutputPCollections(consumer));
+      }
+    } else {
+      // Some of the consumers can't be fused into this subgraph, so the PCollection has to be
```

Comment:
move this block to be inside the body of the for loop and then return instead of break.

This way you don't need the `canFuse` local variable.